### PR TITLE
[Prim]new feature: add register composite rule of ops

### PR DESCRIPTION
**Pull request by cyber-pioneer**
### PR types
New features

### PR changes
Others

### Describe
**New feature**: support register forward composite rule of ops. A non-basic op will be replaced by a set of prim ops.
**step1:**
In blocks of program, search all ops which has been registered composite rule (a function: equal to  origin op, and contains set of primitive ops).
**step2:**
Map args of original op to composite rule.
**step3:**
Replace original op with composite rule. For example: softmax --> {exp, reduce_sum, divide, broadcast}

### Background
We design composite operators in order to minimize the set of primitive operators.  If the program contains only primitive ops, there will be three benefits:
1. It's possible to implement high-order automatic differentiation;
2. It's much easier for AI compiler to commit graph optimization；
3. It's inexpensive for other hardware platform to support PaddlePaddle.

### Design
image

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/autograd/test_composite.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+# limitations under the License.
+
+import numpy as np
+
+import paddle
+import paddle.nn.functional as F
+
+TOLERANCE = {
+    "float32": {
+        "forward": {"rtol": 1e-3, "atol": 1e-3, "eps": 1e-4},
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/unittests/autograd/test_composite.py]: Will this be too large ?

File: python/paddle/incubate/autograd/__init__.py [hunk 2]
```diff
@@ -25,4 +25,5 @@
     'disable_prim',
     'forward_grad',
     'grad',
+    'toprim',
```
Comment (JiabinYang) [inline python/paddle/incubate/autograd/__init__.py]: to_prim

File: python/paddle/incubate/autograd/atomops.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (JiabinYang) [inline python/paddle/incubate/autograd/atomops.py]: primitives.py

File: python/paddle/incubate/autograd/composite_rules.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+from .primreg import REGISTER_COMPOSITE, lookup_composite
+
+
+def _composite(op, *args):
+    _lowerrule = lookup_composite(op.type)
+    return _lowerrule(op, *args)
+
+
+@REGISTER_COMPOSITE('softmax')
+def softmax_composite(op, x, axis):
```
Comment (JiabinYang) [inline python/paddle/incubate/autograd/composite_rules.py]: maybe don't pass op as arguments?

File: python/paddle/incubate/autograd/primapi.py [hunk 5]
```diff
@@ -211,3 +215,20 @@ def grad(outputs, inputs, grad_outputs=None):
     ad.erase_dots(xs_dot)
 
     return xs_bar[0] if isinstance(inputs, framework.Variable) else xs_bar
+
+
+def toprim(blocks):
+    """Search ops with has be registered composite rules and replace them with primitive ops."""
+    if isinstance(blocks, paddle.fluid.framework.Block):
+        if debug:
+            print("origin block=============== ", blocks)
```
Comment (JiabinYang) [inline python/paddle/incubate/autograd/primapi.py]: no print

File: python/paddle/incubate/autograd/primx.py [hunk 6]
```diff
@@ -543,6 +546,122 @@ def expand_nested_list(xs):
...
+
+    ops_to_remove = []
+    vars_to_remove = set()
+
+    # Step2: Process all ops in the target block
+    for op_idx in range(len(block.ops)):
+        op = block.ops[op_idx]
+        ops_to_remove.append(op_idx)
+        if lookup_fn(op.type) is not None and op.type not in blacklist:
+            input_args = get_composite_input_var_list(op)
```
Comment (JiabinYang) [inline python/paddle/incubate/autograd/primx.py]: using `prepare_python_api_arguments` in op_test.py is better

**Review by JiabinYang — COMMENTED** [review]
some coments

File: python/paddle/incubate/autograd/primapi.py (hunk 5 above)
Comment (cyber-pioneer) [reply python/paddle/incubate/autograd/primapi.py]: done

File: python/paddle/incubate/autograd/composite_rules.py (hunk 4 above)
Comment (cyber-pioneer) [reply python/paddle/incubate/autograd/composite_rules.py]: done

File: python/paddle/incubate/autograd/atomops.py (hunk 3 above)
Comment (cyber-pioneer) [reply python/paddle/incubate/autograd/atomops.py]: done

File: python/paddle/incubate/autograd/__init__.py (hunk 2 above)
Comment (cyber-pioneer) [reply python/paddle/incubate/autograd/__init__.py]: done

File: python/paddle/fluid/tests/unittests/autograd/test_composite.py (hunk 1 above)
Comment (cyber-pioneer) [reply python/paddle/fluid/tests/unittests/autograd/test_composite.py]: done

File: python/paddle/incubate/autograd/primx.py (hunk 6 above)
Comment (cyber-pioneer) [reply python/paddle/incubate/autograd/primx.py]: done

File: python/CMakeLists.txt [hunk 7]
```diff
@@ -53,6 +53,28 @@ set(FLUID_CORE_DEPS ${FLUID_CORE})
 
 add_custom_target(copy_libpaddle ALL DEPENDS ${FLUID_CORE_DEPS})
 
+set(ops_yaml_path "${PADDLE_SOURCE_DIR}/paddle/phi/api/yaml/ops.yaml")
```
Comment (JiabinYang) [inline python/CMakeLists.txt:63]: Add todo to move this to prim dir

File: python/paddle/incubate/autograd/composite_rules.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .primitives import *  # noqa: F403
+from .primreg import REGISTER_COMPOSITE, lookup_composite
+
```
Comment (JiabinYang) [inline python/paddle/incubate/autograd/composite_rules.py:22]: Give some comments to show the rules of composite

**Review by JiabinYang — COMMENTED** [review]
two comments

File: python/paddle/incubate/autograd/composite_rules.py (hunk 8 above)
Comment (cyber-pioneer) [reply python/paddle/incubate/autograd/composite_rules.py:22]: done

File: python/CMakeLists.txt (hunk 7 above)
Comment (cyber-pioneer) [reply python/CMakeLists.txt:63]: done

**Review by Ligoml — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: python/paddle/fluid/tests/unittests/composite_ops/test_composite_softmax_grad.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+                'x', shape=inputs.shape, dtype=str(inputs.dtype)
+            )
+            x.stop_gradient = False
+            y = fn(x)
+            blocks = main_program.blocks
+            paddle.incubate.autograd.to_prim(blocks)
+            z = paddle.static.gradients([y], x)
+
+        exe = paddle.static.Executor()
+        exe.run(startup_program)
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/unittests/composite_ops/test_composite_softmax_grad.py:93]: test this with enable_prim?

Comment (JiabinYang) [reply python/paddle/fluid/tests/unittests/composite_ops/test_composite_softmax_grad.py:93]: or disable prim manually and add todo to open it in other pr or test

**Review by JiabinYang — COMMENTED** [review]
two comments

File: python/paddle/fluid/tests/unittests/composite_ops/test_composite_softmax_grad.py (hunk 9 above)
Comment (cyber-pioneer) [reply python/paddle/fluid/tests/unittests/composite_ops/test_composite_softmax_grad.py:93]: will be done in next pr

**Review by JiabinYang — APPROVED** [review]
LGTM

File: paddle/phi/api/yaml/op_compat.yaml [hunk 10]
```diff
@@ -1173,6 +1173,8 @@
 
 - op : softmax
   backward : softmax_grad
+  inputs :
+    x : X
```
Comment (zyfncg) [inline paddle/phi/api/yaml/op_compat.yaml:1177]: outputs不用配置吗？

Comment (cyber-pioneer) [reply paddle/phi/api/yaml/op_compat.yaml:1177]: 我的场景里没有用不用配置

**Review by zyfncg — APPROVED** [review]
